crates/cheatnet/src/execution/syscall_interceptor.rs

```diff
@@ -0,0 +1,206 @@
+use cairo_felt::Felt252;
+use cairo_vm::hint_processor::hint_processor_definition::HintReference;
+use cairo_vm::serde::deserialize_program::ApTracking;
+use cairo_vm::types::exec_scope::ExecutionScopes;
+use cairo_vm::vm::errors::hint_errors::HintError;
+use cairo_vm::vm::errors::vm_errors::VirtualMachineError;
+use cairo_vm::vm::runners::cairo_runner::RunResources;
+use cairo_vm::vm::vm_core::VirtualMachine;
+use std::any::Any;
+use std::collections::HashMap;
+
+pub struct ExecuteHintRequest<'a> {
+    pub vm: &'a mut VirtualMachine,
+    pub exec_scopes: &'a mut ExecutionScopes,
+    pub hint_data: &'a Box<dyn Any>,
+    pub constants: &'a HashMap<String, Felt252>,
+}
+
+pub trait ChainableHintProcessor {
+    fn get_child(&self) -> Option<&dyn HintProcessorLogicInterceptor>;
+    fn get_child_mut(&mut self) -> Option<&mut dyn HintProcessorLogicInterceptor>;
+}
+
+pub trait HintExecutionInterceptor: ChainableHintProcessor {
+    fn intercept_execute_hint(
+        &mut self,
+        execute_hint_request: &mut ExecuteHintRequest,
+    ) -> Option<Result<(), HintError>>;
+
+    fn next_execute_hint(
+        &mut self,
+        execute_hint_request: &mut ExecuteHintRequest,
+    ) -> Option<Result<(), HintError>> {
+        if let Some(result) = self.intercept_execute_hint(execute_hint_request) {
```

Comment:
Can we split execution and interception? Interception should return boolean instead and dont have access to self, to prevent any side effects by mistake. Then if true we should call logic from here.

Reply:
It doesn't make sense to me and would be less maintainable, as we'd have to:
1. Assess the hint type based on the request, and return whether it should be intercepted
2. Match based on the hint type, and redirect to execution (if intercepted)

So hint types would have to be synced in 1 and 2 match arms, which seems like a handful to keep in mind when modifying something